### PR TITLE
Applied arch board feedback for Key Vault

### LINKED ISSUE #18632: Resolve architect board feedback for Key Vault
Here are a few things that came up during the [archboard review](https://github.com/Azure/azure-sdk/issues/2221). Note that there is a discussion about whether to expose a `LocalCryptographyClient` or make it a "mode" of `CryptographyClient`, but that should only apply to .NET and Java. That is not listed below as prototyping is necessary. I will open a new bug shortly.

- [x] Add service version for "7.2". Don't bother with "7.2-preview". It won't be supported anyway, most likely, and "7.2" already is supported by both Key Vault and Managed HSM (rather, Managed HSM ignores it).
- [x] Rename `EncryptOptions` to `EncryptParameters`.
- [x] Rename `DecryptOptions` to `DecryptParameters`.
- [x] `AdditionalAuthenticatedData should be a read-only property (no setters).
- [x] `KeyVaultKeyIdentifier` (as well as for secrets and certificates) should not have "parse" methods, but instead take the URI into the constructor/creator. This pattern will, of course, vary from language to language.
- [x] Make sure all properties on `JsonWebKey` are settable (came up indirectly, but this should be the case already).

### PR DESCRIPTION
**Pull request by vcolin7**
Fixes #18632.

Applied arch board feedback:

- Add service version for `7.2` for KV Keys, Secrets and Certificates.
- Rename `EncryptOptions` to `EncryptParameters` for KV Keys.
- Rename `DecryptOptions` to `DecryptParameters` for KV Keys.
- `KeyVaultKeyIdentifier` (as well as for secrets and certificates) should not have "parse" methods, but instead take the URI into the constructor/creator.
- Make sure all properties on `JsonWebKey` in KV Keys are settable.

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md:3]: Update the release date as well.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateIdentifier.java
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateIdentifier.java:38]: This should throw NPE

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java:262]:
```suggestion
        Objects.requireNonNull(encryptParameters, "'encryptParameters' cannot be null");
```

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java:355]:
```suggestion
        Objects.requireNonNull(decryptParameters, "'decryptParameters' cannot be null");
```

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyServiceClient.java
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyServiceClient.java:129]: Fix all the error messages.
```suggestion
        Objects.requireNonNull(encryptParameters, "'encryptParameters' cannot be null.");
```

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java:59]:
```suggestion
        Objects.requireNonNull(encryptParameters, "'encryptParameters' cannot be null.");
        Objects.requireNonNull(encryptParameters.getAlgorithm(), "encryptParameters.getAlgorithm() cannot be null.");
        Objects.requireNonNull(encryptParameters.getPlainText(), "encryptParameters.getPlainText() cannot be null.");
```

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java:279]: Just "Plaintext". I may have missed this for "Ciphertext" above as well.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java:59]: Just "plaintext". Same with "ciphertext" everywhere.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java:103]: Instances of wrong type/"cipher text" here as well.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java:28]: Just so I make sure it's called out, where is this generated if not specified only for AES-CBC(PAD)? Note that it *must* be a cryptographic RNG.

Comment (heaths) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java:28]: I found it in the `SymmetricKeyCryptographicClient`, but still want to keep this comment open just so that you make sure it's correctly generated only for AES-CBC(PAD) and never AES-GCM.

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md:3]: I prefer to update the release date until I have the exact date, sometimes things get delayed like when waiting for some dependencies to release.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKeyIdentifier.java
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKeyIdentifier.java:47]: Same comment here about asserting - rather, not needing to. IIRC, no other language does because it wasn't really important to. This is just a convenience/helper. There's no harm in, say, parsing a secret ID.

File: sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/models/KeyVaultSecretIdentifier.java
Comment (heaths) [inline sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/models/KeyVaultSecretIdentifier.java:49]: Ditto

**Review by heaths — APPROVED** [review]
I know you're in a crunch, so I'm approving to unblock. But please note there's a number of things that need to be fixed, especially w.r.t. AES-GCM. (Most are minor, but that's a biggie.)

File: eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml
Comment (vcolin7) [reply eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml:2437]: I'm not sure, they've been there for a while. I'll make sure to check if we can get rid of them.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/DecryptParameters.java
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/DecryptParameters.java:50]: We both had a conversation a while ago involving @JonathanGiles about the naming we would follow because some of our APIs were already capitalized using camelCase. In the end we went this route to stay consistent with those older APIs. We can change the capitalization of parameters to be correct, but we cannot change methods, so I think these getters and setters will have to stay that way :/

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateIdentifier.java
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateIdentifier.java:48]: Do you think we should rename the class to `KeyVaultIdentifier` then? Otherwise, I think other languages should verify that the collection is correct, else it could be confusing that a class made specifically for Certificates accepts other collections.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java:43]: Please refer to [my comment](https://github.com/Azure/azure-sdk-for-java/pull/20357#discussion_r607426548) above.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/EncryptParameters.java:279]: Please refer to [my comment](https://github.com/Azure/azure-sdk-for-java/pull/20357#discussion_r607426548) above.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/RsaKeyCryptographyClient.java:59]: Please refer to [my comment](https://github.com/Azure/azure-sdk-for-java/pull/20357#discussion_r607426548) above.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKeyIdentifier.java
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKeyIdentifier.java:47]: Left [a comment](https://github.com/Azure/azure-sdk-for-java/pull/20357#discussion_r607427937) about this above.

File: sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/models/KeyVaultSecretIdentifier.java
Comment (vcolin7) [reply sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/models/KeyVaultSecretIdentifier.java:49]: Left [a comment](https://github.com/Azure/azure-sdk-for-java/pull/20357#discussion_r607427937) about this above.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/DecryptParameters.java
Comment (heaths) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/DecryptParameters.java:77]: Disregard. I found it later and meant to delete this comment, for forgot. Sorry.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/DecryptParameters.java
Comment (heaths) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/DecryptParameters.java:50]: Oh, that's right. Staying consistent with yourselves is better, I agree. Please disregard the other similar comments.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateIdentifier.java
Comment (heaths) [reply sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateIdentifier.java:48]: We wanted separate names to make sure there wasn't any collisions. It's not uncommon to have multiple packages imported simultaneously.

Comment (heaths) [reply sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateIdentifier.java:48]: .NET, JS, and Python don't assert collections (just double checked). Probably best not to here either.

File: eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml
Comment (vcolin7) [reply eng/code-quality-reports/src/main/resources/spotbugs/spotbugs-exclude.xml:2437]: Seems like we don't need that :)

**Review by gapra-msft — APPROVED** [review]

**Pull request merged**
